### ISSUE TITLE
[S-IMPL] Sensitivity Pre-Implementation — Synthetic Only / Primary Freeze Guard

### ISSUE THREAD

**Issue by shota5558** [shota5558/LaggedFacialGraphForecasting#22]
## Objective
Primary の科学的実行順を変更せず、Sensitivity / Optional 解析の **コード実装・契約・synthetic test のみを先行完了**させる。

この Issue では実データによる Sensitivity 結果を生成・閲覧しない。実データでの Sensitivity 実行は #20 `[20][FREEZE] PRIMARY FREEZE` PASS 後、#21 `[21][S] Sensitivity` でのみ行う。

## Depends on
- #4 `[04][CF] Core Contract Freeze`

## Downstream
- #20 `[20][FREEZE] PRIMARY FREEZE`
- #21 `[21][S] Sensitivity`

## Scientific boundary
この Issue で許可するのは **implementation readiness** まで。

### Allowed before Primary Freeze
- source code implementation
- adapter / wrapper implementation
- config schema
- serialization / deserialization
- deterministic seed handling
- synthetic / fixture based unit tests
- synthetic integration tests
- failure / guard tests

### Forbidden before Primary Freeze
- 実データで GPDC を実行して結果を見る
- 実データで LPCMCI を実行して結果を見る
- phase-shuffled / circular-shift surrogate の実データ結果を見る
- `h > 1` の実データ性能を見る
- GRU と Primary Ridge の実データ性能を比較する
- Sensitivity 結果を根拠に Primary config / ParentSet / metric / Null / split / `pc_alpha` / `tau_max` / region definition を変更する
- `artifacts/primary/` を上書きする

原則：
> Sensitivity は先に実装してよいが、科学的実行と結果確認は PRIMARY FREEZE 後に限定する。

---

### S-IMPL-01 Sensitivity execution boundary / namespace
- [ ] Sensitivity 用 config namespace を Primary と分離
- [ ] output root を `artifacts/sensitivity/` に固定
- [ ] Primary artifact を上書きできない guard
- [ ] Primary Freeze 未完了時の real-data Sensitivity execution を拒否する guard
- [ ] synthetic mode と real execution mode を明示的に区別

### S-IMPL-02 PCMCI+ + GPDC adapter
- [ ] Tigramite GPDC dependency / factory
- [ ] PCMCI+ discovery adapter
- [ ] canonical `ParentSet` への変換
- [ ] lagged-link filtering
- [ ] serialization round-trip
- [ ] synthetic known-dependence test
- [ ] deterministic seed / provenance test

**禁止:** Primary Freeze 前に実顔データで結果を生成しない。

### S-IMPL-03 LPCMCI adapter
- [ ] Tigramite LPCMCI adapter
- [ ] config schema
- [ ] canonical graph / parent representation への変換
- [ ] serialization round-trip
- [ ] latent-confounding synthetic fixture
- [ ] deterministic test

**位置づけ:** 未観測共通原因に対する Sensitivity。Primary PCMCI+ + ParCorr を置換しない。

### S-IMPL-04 Phase-shuffled surrogate
- [ ] surrogate generator
- [ ] subject boundary preservation
- [ ] series length / sampling metadata preservation
- [ ] spectrum approximately preserved の検証
- [ ] cross-series temporal alignment destruction の検証
- [ ] deterministic seed test
- [ ] input mutation 禁止 test

### S-IMPL-05 Circular-shift surrogate
- [ ] circular-shift generator
- [ ] sufficiently-large shift policy interface
- [ ] wrapping semantics の明示
- [ ] subject / sequence boundary guard
- [ ] autocorrelation preservation test
- [ ] cross-series alignment destruction test
- [ ] deterministic seed test

### S-IMPL-06 Multi-horizon (`h > 1`) support
- [ ] `DesignMatrix` builder を horizon parameterized にする
- [ ] direct-observation parent availability `tau >= h` guard
- [ ] `lag < h` feature を使用不可にする
- [ ] forecast origin / target time alignment test
- [ ] horizon ごとの provenance 保存
- [ ] synthetic `h=2,3,...` tests

**Scientific constraint:** Primary は `h = 1` のまま。`h > 1` は別 estimand の Sensitivity としてのみ扱う。

### S-IMPL-07 GRU forecaster adapter
- [ ] existing PyTorch implementation を利用した wrapper
- [ ] Primary と同じ `DesignMatrix` / prediction contract に接続
- [ ] train / validation / test scope guard
- [ ] deterministic seed handling
- [ ] tiny synthetic overfit / smoke test
- [ ] `PredictionArtifact` compatibility test

**Scientific constraint:** GRU は confirmation / model-dependence analysis のみ。Primary Ridge を置換しない。

### S-IMPL-08 Sensitivity result / statistics schema
- [ ] sensitivity method identifier
- [ ] source Primary Freeze manifest reference field
- [ ] outer fold / subject / region / horizon provenance
- [ ] metric schema reuse
- [ ] paired-comparison compatibility
- [ ] independent config / seed / software-version provenance
- [ ] Primary result table と混在しない serialization

### S-IMPL-09 Synthetic sensitivity integration gate
以下を synthetic data / fixture のみで end-to-end に通す。

```text
Synthetic FaceTimeSeries
        ↓
Sensitivity method
        ↓
ParentSet / transformed series / DesignMatrix
        ↓
Forecaster where applicable
        ↓
PredictionArtifact / Result Artifact
        ↓
Sensitivity namespace serialization
```

- [ ] GPDC synthetic integration PASS
- [ ] LPCMCI synthetic integration PASS
- [ ] phase-surrogate integration PASS
- [ ] circular-shift integration PASS
- [ ] `h > 1` integration PASS
- [ ] GRU smoke integration PASS
- [ ] repeated same seed gives deterministic artifacts where algorithmically expected

### S-IMPL-10 Primary Freeze execution barrier
- [ ] Primary Freeze manifest が存在しない real-data run を fail-closed
- [ ] invalid / incomplete freeze manifest を拒否
- [ ] synthetic tests は freeze manifest なしでも実行可能
- [ ] real-data Sensitivity artifact が freeze manifest provenance を必須保持
- [ ] Primary artifact write attempt を拒否

---

## Required architecture

```text
Core Contracts
      │
      ├── GPDC adapter ───────────────┐
      ├── LPCMCI adapter ─────────────┤
      ├── phase surrogate ────────────┤
      ├── circular shift ─────────────┤
      ├── h>1 matrix support ─────────┤
      └── GRU adapter ────────────────┤
                                      ↓
                         Synthetic Integration Gate
                                      ↓
                       IMPLEMENTATION READY / CODE FREEZE
                                      ↓
                         [REAL EXECUTION BLOCKED]
                                      ↓
                           #20 PRIMARY FREEZE PASS
                                      ↓
                           #21 Sensitivity Run
```

## Artifact separation

```text
artifacts/primary/       # immutable Primary namespace
artifacts/sensitivity/   # Sensitivity only
artifacts/setup/         # synthetic / implementation gate evidence
```

Primary Freeze 前の synthetic evidence は `artifacts/setup/` 等の非-Primary・非-real-result namespace に保存し、実データ Sensitivity artifact と明確に区別する。

## Tests / acceptance criteria
- [ ] 全 Sensitivity component に unit test がある
- [ ] synthetic / fixture のみで integration test が PASS
- [ ] Primary h=1 semantics を変更しない
- [ ] Primary PCMCI+ + ParCorr + Ridge semantics を変更しない
- [ ] existing Core Contract compatibility を維持
- [ ] outer-test leakage guard を迂回しない
- [ ] sensitivity outputs cannot overwrite `artifacts/primary/`
- [ ] real-data run before #20 PASS is fail-closed
- [ ] config / seed / software version / provenance が保存される
- [ ] #21 が実装不足なしで real Sensitivity execution を開始できる状態になる

## Definition of Done
Sensitivity の GPDC / LPCMCI / phase-shuffled surrogate / circular-shift surrogate / `h>1` / GRU、および共通 result schema・runner guard が実装され、**synthetic-only の unit / integration evidence が PASS**している。

同時に、PRIMARY FREEZE 前の実データ Sensitivity 実行を構造的に拒否し、Primary artifact integrity が維持されていること。

この Issue の完了は **Sensitivity の科学的実行完了を意味しない**。実データ解析は #20 PASS 後に #21 で行う。

## Branch
`dev`

**Comment by shota5558** [comment]
## Completion audit — PASS

Issue #22 has reached **implementation readiness / synthetic-only code freeze** on `dev`.

### Task evidence
- S-IMPL-01 PASS — Sensitivity namespace / execution boundary
- S-IMPL-02 PASS — PCMCI+ + GPDC adapter
- S-IMPL-03 PASS — LPCMCI adapter / latent-confounding graph evidence
- S-IMPL-04 PASS — phase-shuffled surrogate
- S-IMPL-05 PASS — circular-shift surrogate
- S-IMPL-06 PASS — Sensitivity-only `h > 1` support with `tau >= h` guard
- S-IMPL-07 PASS — PyTorch GRU adapter and CPU smoke workflow
- S-IMPL-08 PASS — Sensitivity result/statistics namespace and provenance schema
- S-IMPL-09 PASS — synthetic E2E gate for GPDC/LPCMCI/phase/circular/h>1/GRU
- S-IMPL-10 PASS — complete/hash-verified PRIMARY FREEZE execution barrier

Per-task evidence is stored under `artifacts/setup/S-IMPL-01/` through `S-IMPL-10/`.

### Final verification
- S-IMPL-09 generic synthetic integration: workflow `34215707345` — success
- S-IMPL-09 CPU PyTorch integration: workflow `34215707352` — success
- S-IMPL-10 hardened barrier regression after contract migration: workflow `34216263343` — success
- final `dev` regression including PASS evidence commit: workflow `34216398527` — success

### Scientific audit
- No real facial-data Sensitivity execution was performed.
- Primary remains PCMCI+ + ParCorr + Ridge, `h=1`.
- `h>1`, GPDC, LPCMCI, phase/circular surrogate, and GRU remain Sensitivity-only.
- outer-test discovery/tuning leakage guards were not bypassed.
- Sensitivity output is confined to `artifacts/sensitivity/`; Primary writes are rejected.
- Real Sensitivity remains fail-closed until `artifacts/primary/freeze_manifest.json` is a complete `PASS` manifest containing all Issue #20 freeze targets with existing files and matching SHA-256 hashes.
- This completion does **not** mean Sensitivity scientific execution is complete. Real-data execution remains reserved for #21 after #20 PRIMARY FREEZE PASS.

**Conclusion: Issue #22 Definition of Done = PASS.**

**Comment by shota5558** [comment]
## Completion re-audit after S-IMPL-10 remediation

Issue #22 was reopened because the original S-IMPL-10 validator was correct in isolation, but several public Sensitivity entrypoints could bypass the common Primary Freeze execution barrier.

### Remediation completed

- `bef72dbe08ea23d9ef852b003bc60d52899f73c7`
  - phase shuffle: execution config + common barrier mandatory
  - circular shift: execution config + common barrier mandatory
  - all h>1 Sensitivity design-matrix builders: execution config + common barrier mandatory
  - direct-call fail-closed regression test added
- `cd95c7bd9cc686a121c461c34b7c5bc91ca9ffb7`
  - GRU fit/predict: execution config + common barrier mandatory
  - canonical Sensitivity artifact writer added
  - exact Primary Freeze provenance and sensitivity namespace confinement enforced
- `84b2700947786241160046c56c0c32589f156a1d`
  - raw `serialize_sensitivity_metric` / `dumps_sensitivity_metric` now require an execution permit
  - real serialization validates the complete/hash-verified Primary Freeze and exact provenance before producing bytes
- `fbbbfae71dfc4ccf8a3532f4f7826478195f35bc`
  - S-IMPL-10 evidence updated with the re-audit and remediation record

### Execution evidence

- surrogate/horizon remediation: workflow `34221900278` — **success**
- GRU/writer normal tests: workflow `34222150577` — **success**
- GRU Torch tests: workflow `34222150533` — **success**
- final hardened code normal tests: workflow `34222669956` — **success**
- final hardened code Torch tests: workflow `34222669837` — **success**
- evidence-only final HEAD normal tests: workflow `34222890570` — **success**

### Scientific/contract re-audit

- real Sensitivity before Primary Freeze: **structurally rejected**
- invalid/incomplete/hash-mismatched Primary Freeze: **rejected**
- phase/circular/h>1/GRU direct public execution bypass: **closed**
- raw result serialization bypass: **closed**
- Sensitivity output into `artifacts/primary`: **rejected**
- synthetic pre-implementation before Primary Freeze: **still allowed**
- Primary h=1 / PCMCI+ + ParCorr / Ridge / Core contracts: **unchanged**
- real facial data used during this remediation: **no**

Re-audit verdict: **PASS**. No remaining Issue #22 blocker identified.
